### ISSUE TITLE
Passing request specific values to a IChatCompletion/ITextCompletion instance

### ISSUE THREAD

**Issue by vbossica** [microsoft/semantic-kernel#3379]
Currently, a `IChatCompletion`/`ITextCompletion` class can't receive request specific properties, for example the userId that is doing the invokation. It's obviously not something that can be save in the `config.json` file (and passed through an `AIRequestSettings` class. IMO should be renamed to `AIFunctionSettings`) and also not in the `ContextVariables()` as it's not passed to the `GetChatCompletionsAsync` method.

Ideally something like:

```
var skFunctions = kernel.ImportSemanticFunctionsFromDirectory(skillDirectory, skillDirectoryNames.Split(','));
var context = new ContextVariables();
var requestVariables = new RequestVariables();
requestVariables.Set("userId", ...);

return await kernel.RunAsync(context, requestVariables, skFunctions[pluginName]);
```

and the `IChatCompletion` would have something like:

```
public async Task<IReadOnlyList<IChatResult>> GetChatCompletionsAsync(ChatHistory chat,
    AIRequestSettings? requestSettings = null,
    RequestVariables requestVariables = null,                                                                              
    CancellationToken cancellationToken = default);
```

**Comment by RogerBarreto** [comment]
@vbossica from the Connectors perspective the input should be payload (prompt) on its final state. This today can be accomplished with our PromptTemplate engine and its templates where the variables are interpolated correctly regarding the prepared prompt.
 
So following you first example, the same can be accomplished using the context variables directly:

```
var skFunctions = kernel.ImportSemanticFunctionsFromDirectory(skillDirectory, skillDirectoryNames.Split(','));
var requestVariables = new ContextVariables();
requestVariables.Set("userId", ...);

return await kernel.RunAsync(requestVariables, skFunctions[pluginName]);
```
Your functions template should use something like: 
```
User id: {{$userId}}
User name: ...
```
Please take a look in: [$input example here](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/KernelSyntaxExamples/Example05_InlineFunctionDefinition.cs)

**Comment by matthewbolanos** [comment]
@vbossica, does @RogerBarreto's recommendation solve your issue? If not, please reopen.

**Comment by vbossica** [comment]
I don't think this solves the issue. Let me take another example, as the user was maybe not the best one.
The AI service called within the ChatCompletion and TextCompletion allows certain values (like a `correlationId`) to be passed as a HTTP header. The `correlationId` can't be defined in a skill configuration, nor in a `contextVariable` because it's not accessible within the ChatCompletion and TextCompletion's methods.

**Comment by matthewbolanos** [comment]
Are you able to share how you'd expect the correlationId to be passed all the way into the HTTP request? The correlationId sounds like something that we might need a separate task to complete unless you think there's a generic way of doing it.

**Comment by RogerBarreto** [comment]
@matthewbolanos This is a important need to capture in our APIs, passing metadata in the runtime and making that available in the Connectors.

Thanks for the feedback @vbossica

**Comment by matthewbolanos** [comment]
@SergeyMenshykh, I think this is related to the work you're doing with Stephen. Adding you to this issue to see if the V1.0 proposal fixes this.

**Comment by matthewbolanos** [comment]
We're adding a data property in the kernel so that developers can put ambient data in it and pass it everywhere the kernel goes (including in the kernel). We believe that you'll be able to pass things like the correlationID into it so that you can get a full view of what's happening. 

This will be included in our RC1 release later today. Please give it a try and let us know if it unblocks you.

**Comment by vbossica** [comment]
@matthewbolanos I'm looking at the changes of RC4 right now and I don't think it covers the use case.

In my mind, the Kernel is supposed to be a singleton, created at startup and shared across requests. Storing request-specific values in a singleton will not work (as the values from on request will be overridden by the ones of another one).

Or am I missing something in the Kernel lifecycle? Is it supposed to be recreated at each request?

**Comment by stephentoub** [comment]
> Or am I missing something in the Kernel lifecycle? Is it supposed to be recreated at each request?

Yes. You _can_ create one and reuse it everywhere if that meets your needs, but the intention is now that it is a lightweight thing you can create for each request... in fact if you use the AddKernel method for adding a kernel to a service container, it adds it as transient.

**Comment by vbossica** [comment]
Ah, the "now" is kind of important.

So my issue is now moot.
